cardano-node/src/Cardano/Node/Configuration/TopologyP2P.hs

```diff
@@ -132,77 +138,120 @@ data LocalRootPeersGroup = LocalRootPeersGroup
+-- 'accessPoints', 'advertise' and 'valency' fields are attached to the same
+-- object.
 instance FromJSON LocalRootPeersGroup where
   parseJSON = withObject "LocalRootPeersGroup" $ \o ->
                 LocalRootPeersGroup
-                  <$> o .: "localRoots"
+                  <$> parseJSON (Object o)
                   <*> o .: "valency"
 
 instance ToJSON LocalRootPeersGroup where
   toJSON lrpg =
     object
-      [ "localRoots" .= localRoots lrpg
-      , "valency"    .= valency lrpg
+      [ "accessPoints" .= rootAccessPoints (localRoots lrpg)
+      , "advertise" .= rootAdvertise (localRoots lrpg)
+      , "valency" .= valency lrpg
       ]
 
 newtype LocalRootPeersGroups = LocalRootPeersGroups
   { groups :: [LocalRootPeersGroup]
   } deriving (Eq, Show)
 
 instance FromJSON LocalRootPeersGroups where
-  parseJSON = withObject "LocalRootPeersGroups" $ \o ->
-                LocalRootPeersGroups
-                  <$> o .: "groups"
+  parseJSON = fmap LocalRootPeersGroups . parseJSONList
 
 instance ToJSON LocalRootPeersGroups where
-  toJSON lrpg =
-    object
-      [ "groups" .= groups lrpg
-      ]
+  toJSON = toJSONList . groups
 
 newtype PublicRootPeers = PublicRootPeers
   { publicRoots :: RootConfig
   } deriving (Eq, Show)
 
 instance FromJSON PublicRootPeers where
-  parseJSON = withObject "PublicRootPeers" $ \o ->
-                PublicRootPeers
-                  <$> o .: "publicRoots"
+  parseJSON = fmap PublicRootPeers . parseJSON
 
 instance ToJSON PublicRootPeers where
-  toJSON prp =
-    object
-      [ "publicRoots" .= publicRoots prp
-      ]
+  toJSON = toJSON . publicRoots
 
 data NetworkTopology = RealNodeTopology !LocalRootPeersGroups ![PublicRootPeers] !UseLedger
   deriving (Eq, Show)
 
 instance FromJSON NetworkTopology where
   parseJSON = withObject "NetworkTopology" $ \o ->
-                RealNodeTopology <$> (o .: "LocalRoots"                                     )
-                                 <*> (o .: "PublicRoots"                                    )
+                RealNodeTopology <$> (o .: "localRoots"                                     )
+                                 <*> (o .: "publicRoots"                                    )
                                  <*> (o .:? "useLedgerAfterSlot" .!= UseLedger DontUseLedger)
 
 instance ToJSON NetworkTopology where
   toJSON top =
     case top of
-      RealNodeTopology lrpg prp ul -> object [ "LocalRoots"         .= lrpg
-                                             , "PublicRoots"        .= prp
+      RealNodeTopology lrpg prp ul -> object [ "localRoots"         .= lrpg
+                                             , "publicRoots"        .= prp
                                              , "useLedgerAfterSlot" .= ul
                                              ]
 
+--
+-- Legacy p2p topology file format
+--
+
+-- | A newtype wrapper which provides legacy 'FromJSON' instances.
+--
+newtype Legacy a = Legacy { getLegacy :: a }
```

Comment:
I'd prefer you handle the legacy file format in the original type's JSON instances. I'd rather not introduce another type.

Reply:
This will be removed in `1.37.0` or so, and it makes it very easy to do so. If we'd preserve the original instances, we'd need to have a newtype wrapper for the new ones, and then when we'd remove the legacy format we'd have more work to do.